lib/philomena/images/query.ex

```diff
@@ -173,24 +173,37 @@ defmodule Philomena.Images.Query do
       |> Enum.map(&String.to_integer/1)
       |> Enum.filter(&(&1 <= 2_147_483_647))
 
-  defp preload_filters([], context), do: Map.put(context, :filters, %{})
+  defp load_filters([], _context), do: {:ok, %{}}
 
-  defp preload_filters(ids, context) do
+  defp load_filters(_ids, %{filter: true}),
+    do: {:error, "Filter queries inside filters are not allowed."}
+
+  defp load_filters(ids, _context) do
     filters =
       Filter
       |> where([f], f.id in ^ids)
       |> Repo.all()
       |> Map.new(&{&1.id, &1})
 
-    Map.put(context, :filters, filters)
+    {:ok, filters}
   end
 
-  defp parse(fields, context, query_string) do
-    context = query_string |> parse_filter_ids() |> preload_filters(context)
+  defp prepare_context(context, query_string) do
+    with {:ok, filters} <- query_string |> parse_filter_ids() |> load_filters(context) do
+      {:ok, Map.merge(context, %{filters: filters})}
+    end
```

Comment:
if there's only one point where the pattern diverges, `case` is usually preferred over `with`
```ex
query_string
|> parse_filter_ids()
|> load_filters(context)
|> case do
  {:ok, filters} ->
     {:ok, Map.put(context, :filters, filters)}

  err ->
     err
end
```

Reply:
I did it like this in case `prepare_context` ever needs to do more work than load filters but I can change it if you want